packages/logging/src/logging-and-telemetry.ts

```diff
@@ -0,0 +1,884 @@
+import redactInfo from 'mongodb-redact';
+import { redactURICredentials } from '@mongosh/history';
+import type {
+  MongoshBus,
+  ApiEventWithArguments,
+  ApiEvent,
+  UseEvent,
+  EvaluateInputEvent,
+  ShowEvent,
+  ConnectEvent,
+  ScriptLoadFileEvent,
+  StartLoadingCliScriptsEvent,
+  StartMongoshReplEvent,
+  GlobalConfigFileLoadEvent,
+  CryptLibrarySkipEvent,
+  CryptLibraryFoundEvent,
+  SnippetsCommandEvent,
+  SnippetsErrorEvent,
+  SnippetsFetchIndexErrorEvent,
+  SnippetsFetchIndexEvent,
+  SnippetsLoadedEvent,
+  SnippetsLoadSnippetEvent,
+  SnippetsNpmDownloadActiveEvent,
+  SnippetsNpmDownloadFailedEvent,
+  SnippetsNpmLookupEvent,
+  SnippetsRunNpmEvent,
+  SnippetsTransformErrorEvent,
+  EditorRunEditCommandEvent,
+  EditorReadVscodeExtensionsDoneEvent,
+  EditorReadVscodeExtensionsFailedEvent,
+  FetchingUpdateMetadataEvent,
+  FetchingUpdateMetadataCompleteEvent,
+  SessionStartedEvent,
+  MongoshBusEventsMap,
+  WriteCustomLogEvent,
+} from '@mongosh/types';
+import { inspect } from 'util';
+import { MongoLogWriter } from 'mongodb-log-writer';
+import { mongoLogId } from 'mongodb-log-writer';
+import type {
+  AnalyticsIdentifyMessage,
+  MongoshAnalytics,
+  MongoshAnalyticsIdentity,
+} from './analytics-helpers';
+import type { ConnectEventMap } from '@mongodb-js/devtools-connect';
+import { hookLogger } from '@mongodb-js/devtools-connect';
+import { MultiSet, toSnakeCase } from './helpers';
+import { Writable } from 'stream';
+
+export class MongoshLoggingAndTelemetry {
+  private log: MongoLogWriter | null;
+
+  private readonly trackingProperties: {
+    mongosh_version: string;
+    session_id: string;
+  };
+  private readonly userTraits: AnalyticsIdentifyMessage['traits'] & {
+    [key: string]: unknown;
+  };
+
+  private pendingLogEvents: CallableFunction[] = [];
+
+  private telemetryAnonymousId: string | undefined;
+  private userId: string | undefined;
+  private isDummyLog = true;
+  private isSetup = false;
+  private hookedExternalListenersLogId: string | undefined = undefined;
+
+  constructor(
+    private readonly bus: MongoshBus,
+    private readonly analytics: MongoshAnalytics,
+    providedTraits: { platform: string } & {
+      [key: string]: unknown;
+    },
+    mongoshVersion: string
+  ) {
+    const dummySink = new Writable({
+      write(chunk, encoding, callback) {
+        callback();
+      },
+    });
+    this.log = new MongoLogWriter('', null, dummySink);
+
+    this.userTraits = {
+      ...providedTraits,
+      session_id: this.log.logId,
+    };
+
+    this.trackingProperties = {
+      mongosh_version: mongoshVersion,
+      session_id: this.log.logId,
+    };
+  }
+
+  public setup(): void {
+    if (this.isSetup) {
+      throw new Error('Setup can only be called once.');
+    }
+    this._setupBusEventListeners();
+    this.isSetup = true;
+  }
```

Comment:
As one step here, how about we:

- Define the interface of this class as a separate TS `interface`
- Make sure the class implements this interface
- Stop exporting the class from this file and instead export a function that creates the instance + calls `.setup()` (as well as export the interface itself)
- Use the interface as that function's return type 

That feels like a better way to get out of the ".setup() must always be called after the ctor because ctors should not have side effects" dilemma – ultimately the "ctors should not have side effects" rule is about communication and managing expectations of readers/maintainers, not necessarily about code structure